src/driver/m3_subset_test_point_generator.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2022 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Generates test points for the subset use case with the m3 strategy."""
+
+from typing import Iterable, List
+
+import numpy as np
+from itertools import combinations
+from wfa_planning_evaluation_framework.data_generators.data_set import DataSet
+from wfa_planning_evaluation_framework.driver.test_point_generator import (
+    TestPointGenerator,
+)
+
+
+class M3SubsetTestPointGenerator(TestPointGenerator):
+    """Generates test points for the subset use case with the m3 strategy.
+
+    In M3, the training points are the frequency histograms of
+    - each of the single publishers,
+    - each of the all-but-one subsets of publishers,
+    - the union set of all publishers,
+    a total of (2p + 1) subsets for training, p being the number of publishers.
+    This class generates the test points that represent the remaining
+    (2^p - 1) - (2p + 1) subsets, for the use case that customers use the
+    planning model to predict the r/f of unobserved subsets.
+
+    When the number of subsets (2^p - 1) - (2p + 1) is too large, randomly
+    sample subsets for testing.
+    """
+
+    def __init__(
+        self,
+        dataset: DataSet,
+        campaign_spend_fractions: np.ndarray,
+        max_num_points: int = 2000,
+        rng: np.random.Generator = np.random.default_rng(0),
+    ):
+        """Constructs an M3SubsetTestPointGenerator.
+
+        Args:
+            dataset:  The DataSet for which test points are to be generated.
+            campaign_spend_fractions:  The campaign_spend / inventory_spend
+                at each publisher, where each inventory_spend is given by
+                dataset._max_spends.
+            max_num_points:  If (2^p - 1) - (2p + 1) > max_num_points, then
+                randomly sample max_num_points subsets.
```

Comment:
-> randomly sample max_num_points subsets from the (2^p-1) - (2p+1) subsuts.

Reply:
Done.